app/models/calculated_values/error_handler.rb

```diff
@@ -0,0 +1,131 @@
+        @custom_field_id = custom_field_id
+        @error_code = error_code
+        @missing_custom_field_ids = missing_custom_field_ids
+      end
+
+      def create_calculated_value_error!(customized)
+        CalculatedValueError.create(
+          customized:,
+          custom_field_id:,
+          error_code:,
+          missing_custom_field_ids:
+        )
+      end
+    end
+
+    def self.handle_calculation_errors(customized:, calculation_errors:, calculation_values:, given_values:, calculated_fields:)
+      new(customized, calculation_errors, calculation_values, given_values, calculated_fields).handle_errors
+    end
+
+    def initialize(customized, calculation_errors, calculation_values, given_values, calculated_fields)
+      @customized = customized
+      @calculation_errors = calculation_errors
+      @calculation_values = calculation_values
+      @given_values = given_values
+      @calculated_fields = calculated_fields
+    end
+
+    def handle_errors
+      return unless should_create_errors?
+
+      error_contexts = build_error_contexts
+      create_error_records!(error_contexts)
+    end
+
+    private
+
+    attr_reader :customized, :calculation_errors, :calculation_values, :given_values, :calculated_fields
+
+    def should_create_errors?
+      customized.is_a?(Project)
+    end
+
+    def build_error_contexts
+      calculation_errors.filter_map do |cf_id, error|
+        build_error_context(cf_id, error)
+      end
+    end
+
+    def build_error_context(cf_id, error)
+      custom_field_id = to_id(cf_id)
+
+      case error
+      in :zero_division
+        ErrorContext.new(custom_field_id:, error_code: "ERROR_MATHEMATICAL")
+      in :missing_value
+        build_missing_value_error_context(custom_field_id)
+      in [:unbound_variable, disabled_fields]
+        ErrorContext.new(custom_field_id:,
+                         error_code: "ERROR_DISABLED_VALUE",
+                         missing_custom_field_ids: disabled_fields.map { to_id(it) })
+      else
+        ErrorContext.new(custom_field_id:, error_code: "ERROR_UNKNOWN")
+      end
+    end
+
+    def build_missing_value_error_context(custom_field_id)
+      missing_values = find_missing_values_for_field(custom_field_id)
+      ErrorContext.new(custom_field_id:, error_code: "ERROR_MISSING_VALUE", missing_custom_field_ids: missing_values)
+    end
+
+    def find_missing_values_for_field(custom_field_id)
+      calculated_field = calculated_fields.find { it.id == custom_field_id }
+      return [] unless calculated_field
+
+      calculated_field.formula_referenced_custom_field_ids & cf_ids_with_missing_values
+    end
+
+    # Returns a list of all custom field ids that could not compute a value.
+    def cf_ids_with_missing_values
+      @cf_ids_with_missing_values ||= given_values.merge(calculation_values)
+                                                  .merge(calculation_errors)
```

Comment:
I'm wondering if there is a case when `calculation_errors` misses keys that are present in `calculation_values` and have non `Numeric` value, so maybe it can be just:

```suggestion
      @cf_ids_with_missing_values ||= given_values.merge(calculation_errors)
```

otherwise merge accepts multiple arguments:


```suggestion
      @cf_ids_with_missing_values ||= given_values.merge(calculation_values, calculation_errors)
```

Reply:
`calculation_errors` seems to be enough it seems ✅